src/event/util/log/all_fields.rs

```diff
@@ -0,0 +1,170 @@
+use super::{Atom, Value};
+use serde::{Serialize, Serializer};
+use std::{
+    collections::{btree_map, BTreeMap},
+    iter, slice,
+};
+
+/// Iterates over all paths in form "a.b[0].c[1]" in alphabetical order
+/// and their corresponding values.
+pub fn all_fields<'a>(
+    fields: &'a BTreeMap<Atom, Value>,
+) -> impl Iterator<Item = (Atom, &'a Value)> + Serialize {
+    FieldsIter::new(fields)
+}
+
+#[derive(Clone)]
+enum LeafIter<'a> {
+    Map(btree_map::Iter<'a, Atom, Value>),
+    Array(iter::Enumerate<slice::Iter<'a, Value>>),
+}
+
+#[derive(Clone)]
+enum Node<'a> {
+    Key(&'a Atom),
+    Index(usize),
+}
+
+#[derive(Clone)]
+struct FieldsIter<'a> {
+    stack: Vec<LeafIter<'a>>,
+    nodes: Vec<Node<'a>>,
+}
```

Comment:
Could you add a high-level comment explaining roughly the algorithm here? The use of `stack` is relatively straightforward for a depth-first traversal, but I'm a little less clear on how we're using `nodes`.

Reply:
The `nodes` name was changed to `path` and a comment was added.